src/zenml/pipelines/base_pipeline.py

```diff
@@ -92,42 +92,92 @@ def __init__(self, *args: BaseStep, **kwargs: BaseStep) -> None:
 
         self.enable_cache = getattr(self, PARAM_ENABLE_CACHE)
         self.pipeline_name = self.__class__.__name__
-        self.__steps = dict()
         logger.info(f"Creating pipeline: {self.pipeline_name}")
         logger.info(
             f'Cache {"enabled" if self.enable_cache else "disabled"} for '
             f"pipeline `{self.pipeline_name}`"
         )
 
-        if args:
+        self.__steps: Dict[str, BaseStep] = {}
+        self._verify_arguments(*args, **kwargs)
+
+    def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
+        """Verifies the initialization args and kwargs of this pipeline.
+
+        This method makes sure that no missing/unexpected arguments or
+        arguments of a wrong type are passed when creating a pipeline. If
+        all arguments are correct, saves the steps to `self.__steps`.
+
+        Args:
+            *steps: The args passed to the init method of this pipeline.
+            **kw_steps: The kwargs passed to the init method of this pipeline.
+
+        Raises:
+            PipelineInterfaceError: If there are too many/few arguments or
+                arguments with a wrong name/type.
+        """
+        input_step_keys = list(self.STEP_SPEC.keys())
+        if len(steps) > len(input_step_keys):
             raise PipelineInterfaceError(
-                "You can only use keyword arguments while you are creating an "
-                "instance of a pipeline."
+                f"Too many input steps for pipeline '{self.pipeline_name}'. "
+                f"This pipeline expects {len(input_step_keys)} step(s) "
+                f"but got {len(steps) + len(kw_steps)}."
             )
 
-        for k, v in kwargs.items():
-            if not isinstance(v, BaseStep):
+        combined_steps = {}
+
+        for i, step in enumerate(steps):
+            if not isinstance(step, BaseStep):
                 raise PipelineInterfaceError(
-                    f"When instantiating a pipeline, you can only pass "
-                    f"in @step like annotated objects. You passed in "
-                    f"`{v}` which is of type `{type(v)}`"
+                    f"Wrong argument type (`{type(step)}`) for positional "
+                    f"argument {i} of pipeline '{self.pipeline_name}'. Only "
+                    f"`@step` decorated functions can be used as arguments "
```

Comment:
Or instances which are subclasses of `BaseStep`. We might want to put that in for advanced users?

Reply:
Agreed I'll add that to the error message